node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,226 @@ async fn fetch_block_worker(
+            fetch_finality_signatures_by_block_header(block_header, ctx).await?
+        {
+            ctx.effect_builder
+                .put_signatures_to_storage(finality_signatures)
+                .await;
+        };
+    }
+}
+
+struct BlockSignaturesCollector(Option<BlockSignatures>);
+
+impl BlockSignaturesCollector {
+    fn new() -> Self {
+        Self(None)
+    }
+
+    fn add(&mut self, signatures: BlockSignatures) {
+        match &mut self.0 {
+            None => {
+                self.0 = Some(signatures);
+            }
+            Some(old_sigs) => {
+                for (pub_key, signature) in signatures.proofs {
+                    old_sigs.insert_proof(pub_key, signature);
+                }
+            }
+        }
+    }
+
+    fn check_if_sufficient(
+        &self,
+        validator_weights: &BTreeMap<PublicKey, U512>,
+        finality_threshold_fraction: Ratio<u64>,
+    ) -> bool {
+        self.0.as_ref().map_or(false, |sigs| {
+            are_signatures_sufficient_for_sync_to_genesis(
+                consensus::check_sufficient_finality_signatures(
+                    validator_weights,
+                    finality_threshold_fraction,
+                    sigs,
+                ),
+            )
+        })
+    }
+
+    fn into_inner(self) -> Option<BlockSignatures> {
+        self.0
+    }
+}
+
+// Returns true if the output from consensus can be interpreted
+// as sufficient finality signatures for the sync to genesis process.
+fn are_signatures_sufficient_for_sync_to_genesis(
+    consensus_verdict: Result<(), FinalitySignatureError>,
+) -> bool {
+    match consensus_verdict {
+        Ok(_) | Err(FinalitySignatureError::TooManySignatures { .. }) => true,
+        Err(_) => false,
+    }
+}
+
+// Fetches the finality signatures from the given peer. In case of timeout, it'll
+// retry up to `retries` times. Other errors interrupt the process immediately.
+async fn fetch_finality_signatures_with_retry(
+    block_hash: BlockHash,
+    peer: NodeId,
+    retries: usize,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<FetchedData<BlockSignatures>, FetcherError<BlockSignatures>> {
+    for _ in 0..retries {
+        let maybe_signatures = ctx
+            .effect_builder
+            .fetch::<BlockSignatures>(block_hash, peer)
+            .await;
+        match maybe_signatures {
+            Ok(result) => return Ok(result),
+            Err(FetcherError::TimedOut { .. }) => continue,
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_header(
+    block_header: BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<Option<BlockSignatures>, Error> {
+    let start = Timestamp::now();
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
```

Comment:
Nice, I forgot we should be doing that. Maybe there are other places where we forgot.

Reply:
There are 4 calls to `get_fully_connected_non_joiner_peers` and just 2 of them are followed by `filter_bad_peers`.

Maybe we should use the tactic you proposed in [other comment](https://github.com/casper-network/casper-node/pull/3125#discussion_r903888018) to prevent forgetting? I.e. have one function like:

```rust
async fn get_fully_connected_non_joiner_peers(ctx: &ChainSyncContext<'_>) -> Vec<NodeId> {
     let mut new_peer_list = ctx
            .effect_builder
            .get_fully_connected_non_joiner_peers()
            .await;
    ctx.filter_bad_peers(&mut new_peer_list);
    new_peer_list
}
```